pkg/rc/replication_controller.go

```diff
@@ -654,23 +654,66 @@ func (rc *replicationController) unschedule(txn *auditingTransaction, rcFields f
 // and add a pod on a new node if the following conditions are met:
 //   1) The ineligible pod is unhealthy OR all of the RC's pods are healthy
 //   2) The RC can acquire a mutation lock on its ID
+//   3) OR the RC is disabled, only on ineligible nodes, and has fewer desired
+//      replicas than current (described more in canNodeTransfer())
 // If the conditions are not met, the function will be called again on the next
 // call of meetDesires() should a node still be ineligible. It returns true
 // when a node transfer occurs
-func (rc *replicationController) attemptNodeTransfer(rcFields fields.RC, current types.PodLocations, ineligible types.NodeName, allocAttempts int) (bool, error) {
+func (rc *replicationController) attemptNodeTransfer(rcFields fields.RC, current types.PodLocations, ineligibles []types.NodeName, allocAttempts int) (bool, error) {
+	ok, unlocker, err := rc.canNodeTransfer(rcFields, current, ineligibles)
+	if err != nil {
+		return false, err
+	} else if !ok {
+		return false, nil
+	}
+	if unlocker != nil {
+		defer unlocker.Unlock()
+	}
+
+	ineligible := ineligibles[0]
+	err = rc.swapNodes(rcFields, current, ineligible, allocAttempts)
+	if err != nil {
+		rc.logger.WithError(err).Errorln("could not swap nodes")
+		alertErr := rc.alerter.Alert(rc.alertInfo(rcFields, err.Error()), alerting.LowUrgency)
+		if alertErr != nil {
+			rc.logger.WithError(alertErr).Errorln("unable to send alert")
+		}
+		return false, err
+	}
+
+	rc.logger.Infof("transferred off %s", ineligible)
+	return true, nil
+}
+
+func (rc *replicationController) canNodeTransfer(rcFields fields.RC, current types.PodLocations, ineligibles []types.NodeName) (bool, consul.Unlocker, error) {
+	currentSet := types.NewNodeSet(current.Nodes()...)
+	ineligibleSet := types.NewNodeSet(ineligibles...)
+	if rcFields.Disabled &&
+		currentSet.Equal(ineligibleSet) &&
+		rcFields.ReplicasDesired < len(current) {
+		// If the RC is disabled, all of its nodes are inelgible, and its
+		// desired replica count is greater than its current replica count,
```

Comment:
I think you mean "desired replica count is less than".

Reply:
Yes, thank you.